plugins/vllm/superduper_vllm/model.py

```diff
@@ -1,171 +1,266 @@
+        if async_engine_args.served_model_name is not None:
+            served_model_names = async_engine_args.served_model_name
         else:
-            LLM = _VllmCore
+            served_model_names = [async_engine_args.model]
 
-        self.llm = LLM(**self.vllm_kwargs)
+        model_config = self._async_llm.engine.get_model_config()
 
-    def _generate(self, prompt: str, **kwargs: t.Any) -> str:
-        return self._batch_generate([prompt], **kwargs)[0]
+        if async_engine_args.disable_log_requests:
+            request_logger = None
+        else:
+            from vllm.entrypoints.logger import RequestLogger
+
+            max_log_len = self.vllm_params.get("max_log_len", 100)
+            request_logger = RequestLogger(max_log_len=max_log_len)
+
+        global openai_serving_chat
+        global openai_serving_completion
+
+        response_role = self.vllm_params.get("response_role", "assistant")
+        lora_modules = self.vllm_params.get("lora_modules", None)
+        prompt_adapters = self.vllm_params.get("prompt_adapters", None)
+        chat_template = self.vllm_params.get("chat_template", None)
+        return_tokens_as_token_ids = self.vllm_params.get(
+            "return_tokens_as_token_ids", None
+        )
+
+        from vllm.entrypoints.openai.serving_chat import OpenAIServingChat
+        from vllm.entrypoints.openai.serving_completion import OpenAIServingCompletion
+
+        self.openai_serving_chat = OpenAIServingChat(
+            self._async_llm,
+            model_config,
+            served_model_names,
+            response_role,
+            lora_modules=lora_modules,
+            prompt_adapters=prompt_adapters,
+            request_logger=request_logger,
+            chat_template=chat_template,
+            return_tokens_as_token_ids=return_tokens_as_token_ids,
+        )
+        self.openai_serving_completion = OpenAIServingCompletion(
+            self._async_llm,
+            model_config,
+            served_model_names,
+            lora_modules=lora_modules,
+            prompt_adapters=prompt_adapters,
+            request_logger=request_logger,
+            return_tokens_as_token_ids=return_tokens_as_token_ids,
+        )
+
+    def _chat(
+        self,
+        messages: list["ChatCompletionMessageParam"],
+        **kwargs,
+    ) -> list["RequestOutput"]:
+        self._init_sync_llm()
+        messages = self._preprocess(messages)
+        from vllm import SamplingParams
+
+        sampling_params, kwargs = self._parse_args(SamplingParams, **kwargs)
+        outputs = self._sync_llm.chat(
+            messages=messages,
+            sampling_params=sampling_params,
+            **kwargs,
+        )
+        return self._postprocess_request_output(outputs)
+
+    async def _async_chat(
+        self,
+        messages: list["ChatCompletionMessageParam"],
+        **kwargs,
+    ):
+        messages = self._preprocess(messages)
```

Comment:
This is a sync call? 
if yes it may block the event loop
if its more than few millisecond

Reply:
The post-processing time is negligible.